### PR TITLE
[Feature] Support Fp8 sft in Slime

### PR DESCRIPTION
**Pull request by fy1214**
### Motivation
1.  In order to use the advantage of fp8 to speed up the optimization training process
2. The model trained with fp8 precision has a higher affinity for fp8 reasoning

### Proposed Solution
Adapt FP8 model in Megatron to Slime, and perform the optimal parameter combination under different configurations.

### Implementation Details
1. adapt the environment for transformer_engine fp8 model
2. use blockwise scale weight and activation for fp8 training
3. explore different arguments for the best performance

**bf16 with recompute**
image
Single step time
image
Memory usage
image
**bf16 without recompute**
image
Single step time
image
Memory usage
image
**fp8 with recompute**
image
Single step time
image
Memory usage
image
**fp8 without recompute**
image
Single step time
image
Memory usage
image

4. use activation quantize to reduce the cuda memory usage
the experimental parameter `activation-func-fp8-input-store` is simply implemented with python and triton, but the effect of video memory reduction is obvious.
image
Memory usage
image

### Next Steps / TODO
1. Validate stability improvements during sft training.
2. Validate the accuracy of the training model in different tests.
3. Reduce the use of cuda memory when using large data sets

### CODE REVIEW

**Comment by Baicaihaochi** [comment]
We conducted experiments using the **InfiAlign-72k dataset** (math & code) on the **Qwen2.5-Math-7B-base** model with consistent hyperparameters:
- Batch size: 64
- Learning rate: 2e-5 (min: 2e-6)
- Warm-up ratio: 0.05
- LR decay style: cosine
- Epochs: 5

We compared three training configurations:

| Model                                | AIME25 | AIME24 | MATH500 | GPQA  |
|--------------------------------------|--------|--------|---------|-------|
| InfiAlign-SFT-7B-72k (bf16, OpenRLHF)| 44.27  | 54.70  | 92.60   | 45.33 |
| slime-bf16-7B-72k                    | 44.16  | 56.87  | 92.40   | 45.14 |
| slime-fp8-7B-72k                     | 44.06  | 56.67  | 91.70   | 47.98 |

### Key Observations:
- **slime-fp8** performs nearly on par with **slime-bf16** (and even slightly outperforms on GPQA).
- Both **slime** variants achieve competitive results compared to the **OpenRLHF**-trained baseline.
- The results demonstrate the effectiveness of FP8 precision training in maintaining performance while potentially offering efficiency benefits.

The training loss curves observed on Wandb show normal convergence, indicating stable training behavior:
wandb

**Comment by zhuzilin** [comment]
close as all features have been merged through other PRs, thank you for much for this effort :heart:
